<a href="https://colab.research.google.com/github/hyojunyee/kita_2404/blob/main/m5_%EB%A8%B8%EC%8B%A0%EB%9F%AC%EB%8B%9D/ML_%EC%B6%94%EC%B2%9C_0801.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 추천 알고리즘
크게 세 가지 주요 범주로 나눌 수 있습니다: 콘텐츠 기반 필터링(Content-Based Filtering), 협업 필터링(Collaborative Filtering), 그리고 하이브리드 필터링(Hybrid Filtering)입니다.

콘텐츠 기반 필터링 (Content-Based Filtering)<br>
콘텐츠 기반 필터링은 아이템의 특성(속성)을 기반으로 사용자가 선호할 만한 아이템을 추천하는 방법입니다.
- 특징:
  - 각 아이템의 속성을 벡터로 표현합니다.
  - 사용자의 이전 행동(예: 사용자가 좋아한 아이템)으로부터 프로필을 생성합니다.
  - 사용자가 좋아하는 아이템과 유사한 아이템을 추천합니다.
- 예시:
  - 영화 추천에서, 사용자가 좋아하는 영화의 장르, 감독, 배우 등의 특성을 기반으로 유사한 영화를 추천.
  - 문서 추천에서, 사용자가 읽은 문서의 키워드, 주제 등을 분석하여 유사한 문서를 추천.
- 장점:
  - 새로운 아이템도 쉽게 추천할 수 있습니다(콜드 스타트 문제 해결 가능).
  - 사용자의 개별 취향을 잘 반영합니다.
- 단점:
  - 아이템의 모든 속성을 정의하고 분석하는 것이 어려울 수 있습니다.
  - 사용자가 관심을 보이지 않은 속성은 추천하기 어렵습니다.

협업 필터링 (Collaborative Filtering)<br>
협업 필터링은 사용자와 아이템 간의 상호작용 데이터를 바탕으로 추천을 수행하는 방법입니다. 주로 사용자 간의 유사성 또는 아이템 간의 유사성을 이용합니다.
- 사용자 기반 협업 필터링 (User-Based Collaborative Filtering):
  - 사용자가 유사한 다른 사용자가 좋아한 아이템을 추천합니다.
  - 예: 사용자 A와 B가 유사하다면, B가 좋아한 아이템을 A에게 추천.
- 아이템 기반 협업 필터링 (Item-Based Collaborative Filtering):
  - 사용자가 이전에 좋아한 아이템과 유사한 아이템을 추천합니다.
  - 예: 영화 X와 Y가 유사하다면, X를 본 사용자는 Y도 좋아할 가능성이 높음.
- 잠재요인 협업 필터링 (Latent Factor Collaborative Filtering):
  - 행렬 분해(Matrix Factorization) 기법을 사용하여 사용자와 아이템의 잠재요인을 학습합니다.
  - 예: SVD (Singular Value Decomposition), NMF (Non-negative Matrix Factorization).
- 장점:
  - 아이템의 속성 정보 없이도 추천이 가능합니다.
  - 다양한 사용자 행동 데이터를 활용하여 추천 성능이 좋습니다.
- 단점:
  - 새로운 사용자나 아이템에 대한 정보가 부족한 경우(콜드 스타트 문제) 추천이 어려움.
  - 사용자나 아이템의 수가 많아질수록 계산량이 증가.

하이브리드 필터링 (Hybrid Filtering)<br>
하이브리드 필터링은 콘텐츠 기반 필터링과 협업 필터링을 결합하여 각 접근 방식의 단점을 보완하고 장점을 극대화하는 방법입니다.
- 방법:
  - 두 가지 방법의 결과를 결합하여 최종 추천을 생성합니다.
  - 콘텐츠 기반 추천을 초기 단계에서 사용하고, 이후 협업 필터링을 적용하는 방법.
  - 모델을 결합하여 새로운 하이브리드 모델을 학습하는 방법.
- 장점:
  - 각 방법의 장점을 결합하여 더 정확한 추천을 제공.
  - 콜드 스타트 문제를 완화.
  - 다양한 데이터 소스를 활용하여 추천의 다양성과 정확성 증가.
- 단점:
  - 구현이 복잡하고 계산 비용이 증가할 수 있음.
  - 두 가지 방법의 적절한 조합을 찾기 어려울 수 있음.

## Surprise 패키지
- 파이썬으로 작성된 추천 시스템 라이브러리로, 다양한 추천 알고리즘을 쉽게 사용할 수 있게 도와줍니다.
- Surprise는 특히 행렬 분해(Matrix Factorization)와 같은 협업 필터링 알고리즘을 구현하는 데 강력한 기능을 제공합니다.
- 이 패키지는 사용자-아이템 상호작용 데이터를 기반으로 추천 모델을 구축하고 평가하는 과정을 매우 단순화합니다.

- Surprise 패키지의 주요 기능
  - 다양한 알고리즘 지원: Surprise는 다양한 추천 알고리즘을 제공합니다. 대표적인 알고리즘으로는 다음이 있습니다.
    - 기본 알고리즘: NormalPredictor
    - 협업 필터링: KNNBasic, KNNWithMeans, KNNBaseline
    - 행렬 분해: SVD, SVD++, NMF
    - 베이스라인 알고리즘: BaselineOnly
  - 사용자 정의 데이터셋 지원: Surprise는 내장된 데이터셋 외에도 사용자 정의 데이터셋을 로드할 수 있는 기능을 제공합니다. CSV 파일이나 데이터프레임을 로드하여 사용할 수 있습니다.
  - 모델 평가: Surprise는 다양한 평가 지표를 제공합니다. RMSE, MAE와 같은 지표를 사용하여 모델 성능을 평가할 수 있습니다. 또한, 교차 검증(Cross-validation)과 같은 평가 방법도 지원합니다.
  - 쉽고 직관적인 API: Surprise는 간단하고 직관적인 API를 제공하여 추천 시스템을 쉽게 구현할 수 있도록 도와줍니다.

- Surprise 패키지의 주요 모듈
  - Dataset 모듈:
    - Dataset.load_builtin(name): 내장된 데이터셋을 로드합니다.
    - Dataset.load_from_file(file_path, reader): 파일로부터 데이터셋을 로드합니다.
    - Dataset.load_from_df(df, reader): 데이터프레임으로부터 데이터셋을 로드합니다.
  - Reader 모듈:
    - Reader(line_format, sep, rating_scale): 사용자 정의 데이터셋을 로드할 때 사용되는 클래스입니다.
  - Trainset 클래스:
    - build_full_trainset(): 전체 데이터셋을 학습 데이터로 사용합니다.
    - build_testset(): 전체 데이터셋을 테스트 데이터로 사용합니다.
  - Prediction 모듈:
    - accuracy.rmse(predictions): RMSE를 계산합니다.
    - accuracy.mae(predictions): MAE를 계산합니다.

In [2]:
%pip install scikit-surprise

  Using cached scikit_surprise-1.1.4.tar.gz (154 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  Created wheel for scikit-surprise: filename=scikit_surprise-1.1.4-cp310-cp310-linux_x86_64.whl size=2357251 sha256=3b4246a6203f1deab9444007fd5d93f2413a7e2710426091d48b779103e32146
  Stored in directory: /root/.cache/pip/wheels/4b/3f/df/6acbf0a40397d9bf3ff97f582cc22fb9ce66adde75bc71fd54
Successfully built scikit-surprise


Surprise 패키지에서 제공하는 MovieLens 100k 데이터셋

raw_ratings 속성을 사용하여 데이터셋의 원시 평점 데이터를 가져옵니다. 이 데이터는 사용자 ID, 아이템 ID, 평점, 타임스탬프로 구성된 튜플의 리스트입니다.

In [3]:
import pandas as pd
from surprise import Dataset

# MovieLens 100k 데이터셋 로드
data = Dataset.load_builtin('ml-100k')

# Suprise 데이터셋의 raw_ratings 속성을 사용하여 데이터 확인
raw_ratings = data.raw_ratings

# pandas 데이터프레임으로 변환
df = pd.DataFrame(raw_ratings, columns=['user_id', 'item_id', 'rating', 'timestamp'])

# 데이터프레임 확인
print(df.head())

Dataset ml-100k could not be found. Do you want to download it? [Y/n] Y
Trying to download dataset from https://files.grouplens.org/datasets/movielens/ml-100k.zip...
Done! Dataset ml-100k has been saved to /root/.surprise_data/ml-100k
  user_id item_id  rating  timestamp
0     196     242     3.0  881250949
1     186     302     3.0  891717742
2      22     377     1.0  878887116
3     244      51     2.0  880606923
4     166     346     1.0  886397596


### NormalPredictor:
- 사용자나 아이템의 특성을 고려하지 않고, 단순히 평점의 분포를 기반으로 임의의 예측을 수행합니다. 평점의 평균과 표준편차를 사용하여 임의의 예측 값을 생성합니다.
- 실제 추천 시스템에서는 잘 사용되지 않지만, 비교 기준으로 사용할 수 있습니다.
- 데이터 형태: 사용자 ID, 아이템 ID, 평점, 타임스탬프
- 작동 방식: 주어진 데이터의 평균과 표준편차를 기반으로 무작위 예측을 수행합니다.

In [25]:
from surprise import Dataset, NormalPredictor
from surprise.model_selection import train_test_split
from surprise import accuracy

# 데이터 로드 및 준비
data = Dataset.load_builtin('ml-100k')
trainset, testset = train_test_split(data, test_size=0.25)

# NormalPredictor 모델 사용
algo_NormalPredictor = NormalPredictor()
algo_NormalPredictor.fit(trainset)

# 예측 수행 및 평가, 출력
predictions = algo_NormalPredictor.test(testset)
rmse = accuracy.rmse(predictions, verbose=False)           # 학습 및 평가 과정에서 로그 메세지가 같이 출력 됨. verbose=False는 출력을 없에는 것
print(f"NormalPredictorRMSE: {rmse}")

NormalPredictorRMSE: 1.5199236688812616


### KNNBasic:
- 사용자 기반 또는 아이템 기반의 최근접 이웃 협업 필터링을 수행합니다. 사용자가 유사한 사용자 또는 유사한 아이템을 찾고, 그들의 평점을 기반으로 추천합니다.
- 데이터 형태: 사용자 ID, 아이템 ID, 평점, 타임스탬프
- 작동 방식: 유사도를 계산하여 최근접 이웃을 찾고, 이웃의 평점을 기반으로 예측합니다.

In [21]:
from surprise import KNNBasic

# KNNbasic 모델 사용
algo_knnbasic = KNNBasic()
algo_knnbasic.fit(trainset)

# 예측 수행 및 평가
predictions_knnbasic = algo_knnbasic.test(testset)
rmse = accuracy.rmse(predictions_knnbasic, verbose=False)
print(f"KNNBasic RMSE: {rmse}")

Computing the msd similarity matrix...
Done computing similarity matrix.
KNNBasic RMSE: 0.9813266126594734


### KNNWithMeans
- KNNWithMeans는 KNNBasic과 유사하지만, 각 사용자의 평균 평점을 고려하여 평점을 예측합니다.
- 데이터 형태: 사용자 ID, 아이템 ID, 평점, 타임스탬프
- 작동 방식: 유사도를 계산하여 최근접 이웃을 찾고, 이웃의 평점과 평균 평점을 사용하여 예측합니다.

In [22]:
from surprise import KNNWithMeans

# KNNWithMeans 모델 사용
algo_knnwithmeans = KNNWithMeans()
algo_knnwithmeans.fit(trainset)

# 예측 수행 및 평가
predictions_knnwithmeans = algo_knnwithmeans.test(testset)
rmse = accuracy.rmse(predictions_knnwithmeans, verbose=False)
print(f"KNNWithMeans RMSE: {rmse}")

Computing the msd similarity matrix...
Done computing similarity matrix.
KNNWithMeans RMSE: 0.9444490425881401


SVD (Singular Value Decomposition):
- SVD는 행렬 분해 기반의 협업 필터링 알고리즘입니다. 사용자-아이템 평점 행렬을 분해하여 잠재 요인을 추출하고 이를 기반으로 평점을 예측합니다.
- 데이터 형태: 사용자 ID, 아이템 ID, 평점, 타임스탬프
- 작동 방식: 행렬 분해를 통해 사용자와 아이템의 잠재 요인을 학습하고, 이를 사용하여 평점을 예측합니다.

In [26]:
from surprise import SVD

# SVD 모델 사용
algo_svd = SVD()
algo_svd.fit(trainset)

# 예측 수행 및 평가
predictions_svd = algo_svd.test(testset)
rmse = accuracy.rmse(predictions_svd, verbose=False)
print(f"SVD RMSE: {rmse}")

SVD RMSE: 0.9413906820775302


### NMF (Non-negative Matrix Factorization):
- 비음수 행렬 분해를 사용하는 알고리즘으로, 사용자와 아이템의 잠재 요인을 추출합니다.
- 데이터 형태: 사용자 ID, 아이템 ID, 평점, 타임스탬프
- 작동 방식: 행렬 분해를 통해 비음수 잠재 요인을 학습하고, 이를 사용하여 평점을 예측합니다.

In [24]:
from surprise import NMF

# NMF 모델 사용
algo_nmf = NMF()
algo_nmf.fit(trainset)

# 예측 수행 및 평가
predictions_nmf = algo_nmf.test(testset)
rmse = accuracy.rmse(predictions_nmf, verbose=False)
print(f"NMF RMSE: {rmse}")

NMF RMSE: 0.9592295505644384


KNNbasic 사용하여 추천시스템구현

In [36]:
from surprise import Dataset
from surprise.model_selection import train_test_split

# 데이터 로드 및 준비
data = Dataset.load_builtin('ml-100k')
trainset, testset = train_test_split(data, test_size=0.25)

KNNBasic 알고리즘을 사용하여 추천 모델 훈련

In [44]:
from surprise import KNNBasic
from surprise import accuracy

# KNNbasic 모델 사용
algo = KNNBasic()           # 디폴트는 user_bassed=True -> 유사한 사용자를 찾아 이들의 평점을 기반으로 예측 수행
                            # user_based=False -> 유사한 아이템을 찾아 사용자가 해당 아이템에 매긴 평점을 기반으로 예측
algo.fit(trainset)

# 예측 수행 및 평가
predictions_knnbasic = algo.test(testset)
rmse = accuracy.rmse(predictions_knnbasic)

Computing the msd similarity matrix...
Done computing similarity matrix.
RMSE: 0.9777


특정 사용자에게 추천할 영화를 생성합니다. 예를 들어, 사용자 ID가 196인 경우

In [48]:
# 모든 영화에 대해 예측
user_id = str(15)
items = trainset.all_items()
inner_id_list = [iid for iid in items]
raw_id_list = [trainset.to_raw_iid(iid) for iid in inner_id_list]

predictions = [algo.predict(user_id, raw_id) for raw_id in raw_id_list]

# 예측된 평점 순으로 정렬
predictions.sort(key=lambda x: x.est, reverse=True)

# 상위 10개의 영화 추천
top_n = 20
for prediction in predictions[:top_n]:
    print(f"Movie ID: {prediction.iid}, Predicted Rating: {prediction.est}")

Movie ID: 1293, Predicted Rating: 5
Movie ID: 1358, Predicted Rating: 5
Movie ID: 1189, Predicted Rating: 5
Movie ID: 814, Predicted Rating: 5
Movie ID: 1467, Predicted Rating: 5
Movie ID: 1201, Predicted Rating: 5
Movie ID: 119, Predicted Rating: 5
Movie ID: 1500, Predicted Rating: 5
Movie ID: 1306, Predicted Rating: 5
Movie ID: 1491, Predicted Rating: 5
Movie ID: 850, Predicted Rating: 5
Movie ID: 1516, Predicted Rating: 5
Movie ID: 1599, Predicted Rating: 5
Movie ID: 1607, Predicted Rating: 5
Movie ID: 1449, Predicted Rating: 4.638113797703776
Movie ID: 1367, Predicted Rating: 4.5998594353988524
Movie ID: 50, Predicted Rating: 4.5941469365131224
Movie ID: 1639, Predicted Rating: 4.570095440084836
Movie ID: 1628, Predicted Rating: 4.563309653155035
Movie ID: 1398, Predicted Rating: 4.527777777777779


SVD 알고리즘 사용

In [49]:
from surprise import SVD
from surprise import accuracy

# SVD 모델 사용
algo = SVD()           # 디폴트는 user_bassed=True -> 유사한 사용자를 찾아 이들의 평점을 기반으로 예측 수행
                            # user_based=False -> 유사한 아이템을 찾아 사용자가 해당 아이템에 매긴 평점을 기반으로 예측
algo.fit(trainset)

# 예측 수행 및 평가
predictions_svd = algo.test(testset)
rmse = accuracy.rmse(predictions_svd)

RMSE: 0.9314


In [50]:
# 모든 영화에 대해 예측
user_id = str(15)
items = trainset.all_items()
inner_id_list = [iid for iid in items]
raw_id_list = [trainset.to_raw_iid(iid) for iid in inner_id_list]

predictions = [algo.predict(user_id, raw_id) for raw_id in raw_id_list]

# 예측된 평점 순으로 정렬
predictions.sort(key=lambda x: x.est, reverse=True)

# 상위 10개의 영화 추천
top_n = 20
for prediction in predictions[:top_n]:
    print(f"Movie ID: {prediction.iid}, Predicted Rating: {prediction.est}")

Movie ID: 181, Predicted Rating: 4.103960846124463
Movie ID: 306, Predicted Rating: 4.013871749903276
Movie ID: 50, Predicted Rating: 4.0049177337279795
Movie ID: 316, Predicted Rating: 3.9528308602127695
Movie ID: 603, Predicted Rating: 3.9300066203648236
Movie ID: 302, Predicted Rating: 3.929306527048623
Movie ID: 315, Predicted Rating: 3.9248146610078516
Movie ID: 318, Predicted Rating: 3.9199880601233277
Movie ID: 251, Predicted Rating: 3.919956330055234
Movie ID: 89, Predicted Rating: 3.8994316010260484
Movie ID: 69, Predicted Rating: 3.881729003926008
Movie ID: 114, Predicted Rating: 3.881000740038038
Movie ID: 165, Predicted Rating: 3.8775004861735516
Movie ID: 272, Predicted Rating: 3.8477322486565204
Movie ID: 172, Predicted Rating: 3.8292106503371985
Movie ID: 174, Predicted Rating: 3.8291874003654525
Movie ID: 96, Predicted Rating: 3.813762883836466
Movie ID: 169, Predicted Rating: 3.8130238141832313
Movie ID: 430, Predicted Rating: 3.793193732206147
Movie ID: 56, Predicted 

SVD를 사용한 영화 추천 시스템 구현

In [51]:
import pandas as pd
from surprise import SVD, Dataset, Reader, accuracy
from surprise.dataset import DatasetAutoFolds
from google.colab import drive

# Google Drive 마운트
drive.mount('/content/drive')

# CSV 파일 경로 설정
ratings_file_path = '/content/drive/MyDrive/kdt_240424/m5_머신러닝/dataset/ml-latest-small/ratings.csv'
movies_file_path = '/content/drive/MyDrive/kdt_240424/m5_머신러닝/dataset/ml-latest-small/movies.csv'

# ratings 데이터 로드
ratings = pd.read_csv(ratings_file_path)

# DatasetAutoFolds 클래스를 ratings.csv 파일 기반으로 생성
reader = Reader(line_format='user item rating timestamp', sep=',', rating_scale=(0.5, 5.0))
data_folds = DatasetAutoFolds(ratings_file='/content/drive/MyDrive/kdt_240424/m5_머신러닝/dataset/ml-latest-small/ratings_noh.csv', reader=reader)

# 전체 데이터를 학습 데이터로 생성
trainset = data_folds.build_full_trainset()
algo = SVD(n_epochs=20, n_factors=50, random_state=0)
algo.fit(trainset)

# movies 데이터 로드
movies = pd.read_csv(movies_file_path)

# userId=9 의 movieId 데이터 추출하여 movieId=42 데이터가 있는지 확인
movieIds = ratings[ratings['userId'] == 9]['movieId']

if movieIds[movieIds == 42].count() == 0:
    print('사용자 아이디 9는 영화 아이디 42의 평점 없음')

print(movies[movies['movieId'] == 42])

uid = str(9)
iid = str(42)

pred = algo.predict(uid, iid, verbose=True)
print(ratings[ratings['userId'] == 9]['movieId'].tolist())

# 사용자가 보지 않은 영화 목록 생성 함수
def get_unseen_surprise(ratings, movies, userId):
    seen_movies = ratings[ratings['userId'] == userId]['movieId'].tolist()
    total_movies = movies['movieId'].tolist()
    unseen_movies = [movie for movie in total_movies if movie not in seen_movies]
    print('평점 매긴 영화수:', len(seen_movies), '추천대상 영화수:', len(unseen_movies), '전체 영화수:', len(total_movies))
    return unseen_movies

unseen_movies = get_unseen_surprise(ratings, movies, 9)

# 영화 추천 함수
def recomm_movie_by_surprise(algo, userId, unseen_movies, top_n=10):
    predictions = [algo.predict(str(userId), str(movieId)) for movieId in unseen_movies]

    def sortkey_est(pred):
        return pred.est

    predictions.sort(key=sortkey_est, reverse=True)
    top_predictions = predictions[:top_n]

    top_movie_ids = [int(pred.iid) for pred in top_predictions]
    top_movie_rating = [pred.est for pred in top_predictions]
    top_movie_titles = movies[movies.movieId.isin(top_movie_ids)]['title']
    top_movie_preds = [(id, title, rating) for id, title, rating in zip(top_movie_ids, top_movie_titles, top_movie_rating)]

    return top_movie_preds

unseen_movies = get_unseen_surprise(ratings, movies, 9)
top_movie_preds = recomm_movie_by_surprise(algo, 9, unseen_movies, top_n=10)

print('##### Top-10 추천 영화 리스트 #####')
for top_movie in top_movie_preds:
    print(top_movie[1], ":", top_movie[2])

Mounted at /content/drive
사용자 아이디 9는 영화 아이디 42의 평점 없음
    movieId                   title              genres
38       42  Dead Presidents (1995)  Action|Crime|Drama
user: 9          item: 42         r_ui = None   est = 3.13   {'was_impossible': False}
[41, 187, 223, 371, 627, 922, 923, 1037, 1095, 1198, 1270, 1674, 1987, 2011, 2012, 2023, 2300, 2877, 2901, 3173, 3328, 3735, 4131, 4558, 4993, 5218, 5378, 5445, 5447, 5451, 5481, 5507, 5841, 5843, 5872, 5890, 5891, 5893, 5902, 5952, 5956, 5962, 5965, 5988, 6001, 6044]
평점 매긴 영화수: 46 추천대상 영화수: 9696 전체 영화수: 9742
평점 매긴 영화수: 46 추천대상 영화수: 9696 전체 영화수: 9742
##### Top-10 추천 영화 리스트 #####
Usual Suspects, The (1995) : 4.306302135700814
Star Wars: Episode IV - A New Hope (1977) : 4.281663842987387
Pulp Fiction (1994) : 4.278152632122759
Silence of the Lambs, The (1991) : 4.226073566460876
Godfather, The (1972) : 4.1918097904381995
Streetcar Named Desire, A (1951) : 4.154746591122657
Star Wars: Episode V - The Empire Strikes Back (1980) : 4.122016128